File: 백준/Gold/17298. 오큰수/오큰수.py
import sys

input=sys.stdin.readline
n=int(input())
a=list(map(int,input().split()))
stack=[]
result=[-1]*n
for i in range(len(a)-1):
    # a[i]가 오큰수의 자격이 있는지
    while stack and stack[-1][0]<a[i]:
        result[stack[-1][1]]=a[i]
        stack.pop()
    stack.append((a[i],i))

#print(*stack)
#print(*result)

while stack:
    if stack[-1][0]<a[-1]:
        result[stack[-1][1]]=a[-1]
    else:
        result[stack[-1][1]]=-1
    stack.pop()

#print("==========")
#print(*stack)
print(*result)

File: 백준/Gold/17298. 오큰수/오큰수.java
//package ctp;

import java.io.*;
import java.util.*;

public class Main {
   
	public static class Node{
		int value;
		int index;
		public Node(int value,int index) {
			this.value=value;
			this.index=index;
		}
	}
	
	public static void main(String[] args) throws Exception {
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw=new BufferedWriter(new OutputStreamWriter(System.out));
		int n=Integer.parseInt(br.readLine());
		StringTokenizer st=new StringTokenizer(br.readLine());
		Stack<Node> stack=new Stack<>();
		int[] arr=new int[n];
		int[] result=new int[n];
		for(int i = 0;i<n;i++) {
			result[i]=-1;
		}
		for(int i=0;i<n;i++) {
			arr[i]=Integer.parseInt(st.nextToken());
		}
		for(int i = 0;i<n-1;i++) {
			while(!stack.isEmpty() && stack.peek().value<arr[i]) {
				result[stack.peek().index]=arr[i];
				stack.pop();
			}
			stack.push(new Node(arr[i],i));
		}
		while(!stack.isEmpty()) {
			if(stack.peek().value<arr[n-1]) {
				result[stack.peek().index]=arr[n-1];
			}
			else {
				result[stack.peek().index]=-1;
			}
			stack.pop();
		}
		
		for(int i = 0;i<n;i++) {
			bw.write(result[i]+" ");
		}
		bw.write("\n");
		
		br.close();
		bw.flush();
		bw.close();
	}
	
	
}

